plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java

```diff
@@ -0,0 +1,127 @@
+
+    private static final Logger logger = LogManager.getLogger(S3RetryingInputStream.class);
+
+    private final S3BlobStore blobStore;
+    private final String blobKey;
+    private final int maxAttempts;
+
+    private InputStream currentStream;
+    private long currentOffset;
+
+    S3RetryingInputStream(S3BlobStore blobStore, String blobKey) throws IOException {
+        this.blobStore = blobStore;
+        this.blobKey = blobKey;
+        this.maxAttempts = blobStore.getMaxRetries() + 1;
+        currentStream = openStream();
+    }
+
+    private InputStream openStream() throws IOException {
+        try (AmazonS3Reference clientReference = blobStore.clientReference()) {
+            final GetObjectRequest getObjectRequest = new GetObjectRequest(blobStore.bucket(), blobKey);
+            if (currentOffset > 0) {
+                getObjectRequest.setRange(currentOffset);
+            }
+            final S3Object s3Object = SocketAccess.doPrivileged(() -> clientReference.client().getObject(getObjectRequest));
+            return s3Object.getObjectContent();
+        } catch (final AmazonClientException e) {
+            if (e instanceof AmazonS3Exception) {
+                if (404 == ((AmazonS3Exception) e).getStatusCode()) {
+                    throw new NoSuchFileException("Blob object [" + blobKey + "] not found: " + e.getMessage());
+                }
+            }
+            throw e;
+        }
+    }
+
+    @Override
+    public int read() throws IOException {
+        int attempt = 0;
+        while (true) {
+            attempt += 1;
+            try {
+                final int result = currentStream.read();
+                currentOffset += 1;
+                return result;
+            } catch (IOException e) {
+                reopenStreamOrFail(attempt, e);
+            }
+        }
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        int attempt = 0;
+        while (true) {
+            attempt += 1;
+            try {
+                final int bytesRead = currentStream.read(b, off, len);
+                if (bytesRead == -1) {
+                    return -1;
+                }
+                currentOffset += bytesRead;
+                return bytesRead;
+            } catch (IOException e) {
+                reopenStreamOrFail(attempt, e);
+            }
+        }
+    }
+
+    private void reopenStreamOrFail(int attempt, IOException e) throws IOException {
+        if (attempt >= maxAttempts) {
+            throw e;
+        }
+        logger.debug(new ParameterizedMessage("failed reading [{}/{}] at offset [{}], attempt [{}] of [{}], retrying",
+            blobStore.bucket(), blobKey, currentOffset, attempt, maxAttempts), e);
+        IOUtils.closeWhileHandlingException(currentStream);
+        currentStream = openStream();
+    }
+
+    @Override
+    public void close() throws IOException {
+        currentStream.close();
```

Comment:
I think we want to maintain an internal `closed` flag and prevent any future reads if the `S3RetryingInputStream` has been closed.

Reply:
Sure; we're now asserting that we don't use the stream after it's closed in d9890c6ee12ecc6c24c6189a7f21e5583dfb248e (and throwing an ISE if assertions are disabled)